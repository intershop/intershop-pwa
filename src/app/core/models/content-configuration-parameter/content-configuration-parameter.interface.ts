export interface ContentConfigurationParameterData {
  definitionQualifiedName: string;
  value: string | object;
}
