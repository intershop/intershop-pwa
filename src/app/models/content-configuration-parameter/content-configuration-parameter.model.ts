export interface ContentConfigurationParameter {
  definitionQualifiedName: string;
  value: string;
}
