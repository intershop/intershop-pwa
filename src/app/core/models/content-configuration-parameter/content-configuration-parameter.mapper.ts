import { Injectable } from '@angular/core';
import { Store, select } from '@ngrx/store';

import { getCurrentLocale, getICMStaticURL } from 'ish-core/store/core/configuration';
import { mapToProperty } from 'ish-core/utils/operators';

import { ContentConfigurationParameterData } from './content-configuration-parameter.interface';

export interface ContentConfigurationParameters {
  [key: string]: string | object | number;
}

@Injectable({ providedIn: 'root' })
export class ContentConfigurationParameterMapper {
  private staticURL: string;
  private lang = '-';

  constructor(store: Store) {
    store.pipe(select(getICMStaticURL)).subscribe(url => (this.staticURL = url));
    store.pipe(select(getCurrentLocale), mapToProperty('lang')).subscribe(lang => (this.lang = lang || '-'));
  }

  fromData(data: { [name: string]: ContentConfigurationParameterData }): ContentConfigurationParameters {
    let configurationParameters = {};

    if (data) {
      configurationParameters = Object.entries(data)
        .map(([key, value]) => ({ [key]: value.value }))
        .reduce((acc, val) => ({ ...acc, ...val }));
    }

    this.postProcessFileReferences(configurationParameters);

    return configurationParameters;
  }

  /**
   * TODO: make this dependant on the type of the configuration parameter once the CMS REST API provides this information
   * for now filter all configuration parameters that start with 'Image' or 'Video'
   * and where the value does not start with 'http' but contains ':/'
   + if the filter matches convert the CMS REST API value in a full server URL to the configured file
   */
  postProcessFileReferences(data: ContentConfigurationParameters): ContentConfigurationParameters {
    Object.keys(data)
      .filter(
        key =>
          (key.startsWith('Image') || key.startsWith('Video')) &&
          data[key] &&
          !data[key].toString().startsWith('http') &&
          data[key].toString().includes(':/')
      )
      .forEach(key => {
        const split = data[key].toString().split(':');
        data[key] = encodeURI(`${this.staticURL}/${split[0]}/${this.lang}${split[1]}`);
      });
    return data;
  }
}
