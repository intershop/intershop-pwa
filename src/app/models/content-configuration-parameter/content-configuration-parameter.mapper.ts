import { ContentConfigurationParameterData } from './content-configuration-parameter.interface';

export declare interface ContentConfigurationParameters {
  [key: string]: string | object;
}

export class ContentConfigurationParameterMapper {
  static fromData(data: { [name: string]: ContentConfigurationParameterData }): ContentConfigurationParameters {
    return !data
      ? {}
      : Object.entries(data)
          .map(([key, value]) => ({ [key]: value.value }))
          .reduce((acc, val) => ({ ...acc, ...val }));
  }
}
