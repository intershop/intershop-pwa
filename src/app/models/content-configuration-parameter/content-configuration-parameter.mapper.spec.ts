import { ContentConfigurationParameterData } from './content-configuration-parameter.interface';
import { ContentConfigurationParameterMapper } from './content-configuration-parameter.mapper';

describe('Content Configuration Parameter Mapper', () => {
  it('should return a value for undefined input', () => {
    const result = ContentConfigurationParameterMapper.fromData(undefined);
    expect(result).not.toBeUndefined();
    expect(result).toBeEmpty();
  });

  it('should map to simple dictionary', () => {
    const input: { [name: string]: ContentConfigurationParameterData } = {
      key1: {
        definitionQualifiedName: 'name1',
        value: '1',
      },
      key2: {
        definitionQualifiedName: 'name2',
        value: 'hello',
      },
      key3: {
        definitionQualifiedName: 'name3',
        value: ['hello', 'world'],
      },
    };

    const result = ContentConfigurationParameterMapper.fromData(input);
    expect(result).not.toBeEmpty();
    expect(result).toHaveProperty('key1', '1');
    expect(result).toHaveProperty('key2', 'hello');
    expect(result).toHaveProperty('key3', ['hello', 'world']);
    expect(result).toMatchSnapshot();
  });
});
