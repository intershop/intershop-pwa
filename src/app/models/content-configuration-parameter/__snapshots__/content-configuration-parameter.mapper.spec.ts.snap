// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Content Configuration Parameter Mapper should map to simple dictionary 1`] = `
Object {
  "key1": "1",
  "key2": "hello",
  "key3": Array [
    "hello",
    "world",
  ],
}
`;
